3.Intentionally introduce the following errors into your code and report your opinion as to whether it's easier to (a) work in a ChatBot session to fix the errors, or (b) use google to search for and fix errors: first share the errors you get in the ChatBot session and see if you can work with ChatBot to troubleshoot and fix the coding errors, and then see if you think a google search for the error provides the necessary toubleshooting help more quickly than ChatGPT

A.Forget to include import pandas as pd in your code
Use Kernel->Restart from the notebook menu to restart the jupyter notebook session unload imported libraries and start over so you can create this error

In [1]:


#Load the titanic data set
url =  " https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

df.groupby("sex")["survived"].describe()

NameError: name 'pd' is not defined

In [2]:
import pandas as pd

#Load the titanic data set
url =  " https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

df.groupby("sex")["survived"].describe()


count      mean       std  min  25%  50%  75%  max
sex                                                       
female  314.0  0.742038  0.438211  0.0  0.0  1.0  1.0  1.0
male    577.0  0.188908  0.391775  0.0  0.0  0.0  0.0  1.0

B.Mistype "titanic.csv" as "titanics.csv"
If ChatBot troubleshooting is based on downloading the file, just replace the whole url with "titanics.csv" and try to troubleshoot the subsequent FileNotFoundError: [Errno 2] No such file or directory: 'titanics.csv' (assuming the file is indeed not present)

Explore introducing typos into a couple other parts of the url and note the slightly different errors this produces

In [3]:
import pandas as pd

#Load the titanic data set
url =  " https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanics.csv"
df = pd.read_csv(url)

df.groupby("sex")["survived"].describe()


HTTPError: HTTP Error 404: Not Found

In [4]:
import pandas as pd

# Load the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

# Group by 'sex' and describe the 'survived' column
df.groupby("sex")["survived"].describe()


count      mean       std  min  25%  50%  75%  max
sex                                                       
female  314.0  0.742038  0.438211  0.0  0.0  1.0  1.0  1.0
male    577.0  0.188908  0.391775  0.0  0.0  0.0  0.0  1.0

Introducing error in seaborn-data by removing the dash.

In [5]:
import pandas as pd

# Load the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborndata/master/titanic.csv"
df = pd.read_csv(url)

# Group by 'sex' and describe the 'survived' column
df.groupby("sex")["survived"].describe()

HTTPError: HTTP Error 404: Not Found

In [6]:
import pandas as pd

# Load the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

# Group by 'sex' and describe the 'survived' column
df.groupby("sex")["survived"].describe()


count      mean       std  min  25%  50%  75%  max
sex                                                       
female  314.0  0.742038  0.438211  0.0  0.0  1.0  1.0  1.0
male    577.0  0.188908  0.391775  0.0  0.0  0.0  0.0  1.0

Introducing typo in https by making is httpss

In [7]:
import pandas as pd

# Load the Titanic dataset
url = "httpss://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

# Group by 'sex' and describe the 'survived' column
df.groupby("sex")["survived"].describe()


ValueError: Protocol not known: httpss

Introcuing type in url by changing raw.githubusercontent to raww.githubusercontent

In [8]:
import pandas as pd

# Load the Titanic dataset
url = "httpss://raww.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

# Group by 'sex' and describe the 'survived' column
df.groupby("sex")["survived"].describe()


ValueError: Protocol not known: httpss

In conclusion for the typos I introduced I got three different errors 
-NameError : Occurs when python can't comprehend what name you are trying to use for example when I didn't import pandas pd python could not comprehend what pd.read_csv(url) specifically pd was because pandas was not imported.
-ValueError: When the function you provide python with doesn't have the correct value type so for ex string int, therefore, in my URL when i mistyped https as httpss the pd.read_csv(url)
-HTTPError: Occurs when you are trying to access a url that doesn't exist - the url does exists just like in the case where I mistyped titanic as titanics however the file which i'm trying to access is not found because of the type the incorrect file name.

C.Try to use a dataframe before it's been assigned into the variable
You can simulate this by just misnaming the variable. For example, if you should write df.groupby("col1")["col2"].describe() based on how you loaded the data, then instead write DF.groupby("col1")["col2"].describe()

In [11]:
import pandas as pd

# Load the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

# Group by 'sex' and describe the 'survived' column
DF.groupby("sex")["survived"].describe()

NameError: name 'DF' is not defined

In [13]:
import pandas as pd

# Load the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

# Group by 'sex' and describe the 'survived' column
df.groupby("sex")["survived"].describe()


count      mean       std  min  25%  50%  75%  max
sex                                                       
female  314.0  0.742038  0.438211  0.0  0.0  1.0  1.0  1.0
male    577.0  0.188908  0.391775  0.0  0.0  0.0  0.0  1.0

D.Forget one of the parentheses somewhere the code
For example, if the code should be pd.read_csv(url) the change it to pd.read_csv(url

In [14]:
import pandas as pd

# Load the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

# Group by 'sex' and describe the 'survived' column
df.groupby("sex"["survived"].describe()

SyntaxError: incomplete input (641649791.py, line 8)

In [15]:
import pandas as pd

# Load the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

# Group by 'sex' and describe the 'survived' column
df.groupby("sex")["survived"].describe()

count      mean       std  min  25%  50%  75%  max
sex                                                       
female  314.0  0.742038  0.438211  0.0  0.0  1.0  1.0  1.0
male    577.0  0.188908  0.391775  0.0  0.0  0.0  0.0  1.0

E.Mistype one of the names of the chained functions with the code
For example, try something like df.group_by("col1")["col2"].describe() and df.groupby("col1")["col2"].describle()


In [16]:
import pandas as pd

# Load the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

# Group by 'sex' and describe the 'survived' column
df.grouppby("sex")["survived"].describe()

AttributeError: 'DataFrame' object has no attribute 'grouppby'

In [17]:
import pandas as pd

# Load the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

# Group by 'sex' and describe the 'survived' column
df.groupby("sex")["survived"].describle()

AttributeError: 'SeriesGroupBy' object has no attribute 'describle'

F.Use a column name that's not in your data for the groupby and column selection
For example, try capitalizing the columns for example replacing "sex" with "Sex" in titanic_df.groupby("sex")["age"].describe(), and then instead introducing the same error of "age"

In [18]:
import pandas as pd

# Load the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

# Group by 'sex' and describe the 'survived' column
df.groupby("sex")["SUrvived"].describe()

KeyError: 'Column not found: SUrvived'

G.Forget to put the column name as a string in quotes for the groupby and column selection, and see if the ChatBot and google are still as helpful as they were for the previous question
For example, something like titanic_df.groupby(sex)["age"].describe(), and then titanic_df.groupby("sex")[age].describe()

In [19]:
import pandas as pd

# Load the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)

# Group by 'sex' and describe the 'survived' column
df.groupby("sex")[survived].describe()

NameError: name 'survived' is not defined

Opinion: From my experiments from question number 8.3 I have learnt that chatgpt works was faster and is way more helpful when it comes to fixing errors in python because a google search provides a lot of examples and a lot of cases wheras chapgpt gets straight to the point you also have the oppurtunity of sharing your code with chatgpt for further assistance making it much faster. Whereas on a google search you most likely have to click through different links just to find one piece of information. 

# Summary of Chatbot Session

Certainly! Here’s a summary of our chat session:

Errors Encountered:

NameError: This occurred when using a variable or function name that Python didn’t recognize, such as pd before importing pandas or DF when df was defined.
ValueError: This happened due to incorrect input values, like using httpss instead of https in a URL.
HTTPError 404: This error indicated that the URL you tried to access did not exist.
SyntaxError: This was caused by missing or misplaced parentheses in the code.
AttributeError: This was due to a typo (grouppby instead of groupby).
KeyError: This was due to using the wrong column name case ("SUrvived" instead of "survived").
NameError (repeated): This occurred when using a column name without quotes (survived instead of "survived").
Solutions Provided:

Import pandas correctly using import pandas as pd.
Correct URL typos to avoid HTTPError 404.
Fix syntax errors by ensuring all parentheses are properly closed.
Correct typos like grouppby to groupby.
Check and use correct column names (case-sensitive) by printing column names using df.columns.
Use quotes around column names when referring to them as strings.
If you have any more questions or need further assistance, feel free to ask!

url: https://chatgpt.com/share/5b1409d8-0f25-41aa-8caf-7d0e4938e011